cnd/src/btsieve/bitcoin/bitcoin_cache.rs

```diff
@@ -0,0 +1,94 @@
+use crate::btsieve::{BlockByHash, LatestBlock};
+use bitcoin::{hashes::sha256d, util::hash::BitcoinHash};
+use fmt::Debug;
+use futures::Future;
+use futures_core::{
+    compat::Future01CompatExt,
+    future::{FutureExt, TryFutureExt},
+};
+use std::{fmt, fmt::Formatter, sync::Arc};
+use tokio::sync::Mutex;
+
+pub struct BitcoinCache<T> {
+    pub inner: T,
+    pub cache: Arc<Mutex<lru::LruCache<sha256d::Hash, bitcoin::Block>>>,
+}
+
+impl<T> Clone for BitcoinCache<T>
+where
+    T: Clone,
+{
+    fn clone(&self) -> Self {
+        BitcoinCache {
+            inner: self.inner.clone(),
+            cache: self.cache.clone(),
+        }
+    }
+}
+
+impl<T> Debug for BitcoinCache<T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
+        self.inner.fmt(f)
+    }
+}
+
+impl<T> LatestBlock for BitcoinCache<T>
+where
+    T: LatestBlock<Block = bitcoin::Block, BlockHash = sha256d::Hash> + Clone,
+{
+    type Block = bitcoin::Block;
+    type BlockHash = sha256d::Hash;
+
+    fn latest_block(
+        &mut self,
+    ) -> Box<dyn Future<Item = Self::Block, Error = anyhow::Error> + Send + 'static> {
+        let cache = self.cache.clone();
+        let mut inner = self.inner.clone();
+
+        let future = async move {
+            let block = inner.latest_block().compat().await?;
+            let mut guard = cache.lock().await;
+            guard.put(block.bitcoin_hash(), block.clone());
+            Ok(block)
+        }
+        .boxed()
+        .compat();
+
+        Box::new(future)
+    }
+}
+
+impl<T> BlockByHash for BitcoinCache<T>
```

Comment:
We may want to convert these traits to new futures as a first step to this, it would save us converting all the new code.  @thomaseizinger is there any reason these have _not_ been done before (before @da-kami goes on a wild goose chase)?

Reply:
See https://github.com/comit-network/comit-rs/issues/1879 plus dependency chain.